core/pkg/monitor/gpu_nvidia.go

```diff
@@ -237,71 +178,77 @@ func (g *GPUNvidia) ClearMetrics() {
 	g.mutex.Lock()
 	defer g.mutex.Unlock()
 
-	g.metrics = map[string][]float64{}
+	g.metrics = map[string][]any{}
 }
 
 func (g *GPUNvidia) IsAvailable() bool {
-	g.mutex.Lock()
-	defer g.mutex.Unlock()
-
-	defer func() {
-		if r := recover(); r != nil {
-			g.nvmlInit = nvml.ERROR_UNINITIALIZED
-		}
-	}()
-	g.nvmlInit = nvml.Init()
-	return g.nvmlInit == nvml.SUCCESS
+	return isRunning(g.cmd)
 }
 
 func (g *GPUNvidia) Close() {
-	g.mutex.Lock()
-	defer g.mutex.Unlock()
-
-	if g.nvmlInit != nvml.SUCCESS {
-		return
-	}
-
-	err := nvml.Shutdown()
-	if err != nvml.SUCCESS {
-		return
+	// send signal to close
+	if isRunning(g.cmd) {
+		if err := g.cmd.Process.Signal(os.Kill); err != nil {
+			return
+		}
 	}
 }
 
 func (g *GPUNvidia) Probe() *service.MetadataRequest {
-	g.mutex.Lock()
-	defer g.mutex.Unlock()
-
-	if g.nvmlInit != nvml.SUCCESS {
+	if !g.IsAvailable() {
 		return nil
 	}
 
+	// wait for the first sample
+	for {
+		g.mutex.RLock()
+		_, ok := g.sample["_gpu.count"]
+		g.mutex.RUnlock()
+		if ok {
+			break
+		}
+		// sleep for a while
+		time.Sleep(100 * time.Millisecond)
+	}
+
 	info := service.MetadataRequest{
 		GpuNvidia: []*service.GpuNvidiaInfo{},
 	}
 
-	count, ret := nvml.DeviceGetCount()
-	if ret != nvml.SUCCESS {
+	if count, ok := g.sample["_gpu.count"].(float64); ok {
+		info.GpuCount = uint32(count)
+	} else {
+		return nil
+	}
+
+	// no GPU found, so close the GPU monitor
+	if info.GpuCount == 0 {
+		g.Close()
 		return nil
 	}
 
-	info.GpuCount = uint32(count)
-	names := make([]string, count)
+	if v, ok := g.sample["cuda_version"]; ok {
+		info.CudaVersion = v.(string)
```

Comment:
this risks a panic if the subprocess misbehaves. We should check the version. Better yet, how about unmarshalling into a struct?

Reply:
I structured the output of the external command in a way that can be immediately passed to the backend. let me see if i can come up with a nice intermediate struct for this:
```
{"_timestamp":1722902546.2584937,"cuda_version":"12.4","gpu.0.architecture":"Turing","gpu.0.brand":"Nvidia","gpu.0.cudaCores":2560,"gpu.0.encoderUtilization":0,"gpu.0.enforcedPowerLimitWatts":70.0,"gpu.0.gpu":0,"gpu.0.graphicsClock":300,"gpu.0.maxPcieLinkGen":3,"gpu.0.maxPcieLinkWidth":16,"gpu.0.memory":0,"gpu.0.memoryAllocated":2.8804524739583335,"gpu.0.memoryAllocatedBytes":463929344,"gpu.0.memoryClock":405,"gpu.0.memoryTotal":16106127360,"gpu.0.name":"Tesla T4","gpu.0.pcieLinkGen":1,"gpu.0.pcieLinkSpeed":2500000000,"gpu.0.pcieLinkWidth":16,"gpu.0.powerPercent":13.518571428571427,"gpu.0.powerWatts":9.463,"gpu.0.temp":40,"gpu.1.architecture":"Turing","gpu.1.brand":"Nvidia","gpu.1.cudaCores":2560,"gpu.1.encoderUtilization":0,"gpu.1.enforcedPowerLimitWatts":70.0,"gpu.1.gpu":0,"gpu.1.graphicsClock":300,"gpu.1.maxPcieLinkGen":3,"gpu.1.maxPcieLinkWidth":16,"gpu.1.memory":0,"gpu.1.memoryAllocated":2.8804524739583335,"gpu.1.memoryAllocatedBytes":463929344,"gpu.1.memoryClock":405,"gpu.1.memoryTotal":16106127360,"gpu.1.name":"Tesla T4","gpu.1.pcieLinkGen":1,"gpu.1.pcieLinkSpeed":2500000000,"gpu.1.pcieLinkWidth":16,"gpu.1.powerPercent":19.47,"gpu.1.powerWatts":13.629,"gpu.1.temp":48,"gpu.count":2}
```

Reply:
i experimented with using a struct here: https://github.com/wandb/wandb/pull/8091/files#diff-f8a7efb5bd6b1f9f4006b40ffb89e227309a25fee1795d1925516051c04f418fR53
but so far it gets really ugly, so I'll defer it to after I do the clean up promised in https://wandb.atlassian.net/browse/WB-20288